### ISSUE TITLE
Partial Caching Parsing Error - <% base_tag %> treated as a control block

### ISSUE THREAD

**Issue by HARVS1789UK** [silverstripe/silverstripe-framework#5150]
This can be seen in a fresh install of 3.3.1 (and as far back as 3.1.x if not further).

With Silverstripe partial caching it is recommended that you use a low specificity cache key to cache most things (i.e. the entire page content) and use nested cache blocks/keys with a higher level of specificity to ensure that more complicated page sections do not become stale. It is also made clear that you cannot nest ```<% cached %>``` blocks within ```<% if %>```, ```<% with %>```, ```<% loop %>``` or other 'control blocks'.

Doing so will return you a user error in dev mode) or a server error (in live mode). There is a bug which can be seen detailed in this [forum post](http://www.silverstripe.org/community/forums/general-questions/show/101584)) where if you include the ```<% base_tag %>``` template logic/variable in a ```<% cached %>``` block, any additional nested ```<% cached %>``` or ```<% uncached %>``` blocks throw this same error (presumably as the parser believes they are within an open control block, because ```<% base_tag %>```
![partial-caching-parse-error](https://cloud.githubusercontent.com/assets/5114645/13568216/b7be6d12-e457-11e5-8048-b007cb744e31.png)
![partial-caching-parse-error](https://cloud.githubusercontent.com/assets/5114645/13568222/bf0cb9a2-e457-11e5-861d-80af35b86abc.png)

 has no closing element).

The below example therefore (incorrectly in my opinion) causes a parse error:

```html
<% cached 'Head', $ID, $LastEdited %>
<head>

	<% base_tag %>

	<% cached 'Meta', $SiteConfig.ID, $SiteConfig.LastEdited %>
	<title>
		<% if $MetaTitle %>
			$MetaTitle
		<% else %>
			$Title
		<% end_if %>
		&raquo; $SiteConfig.Title
	</title>
	$MetaTags(false)
	<% end_cached %>

	<meta charset="utf-8">
	<meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0">
	<meta http-equiv="Content-Type" content="text/html; charset=utf-8">

	<!--[if lt IE 9]>
	<script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
	<![endif]-->
	<% require themedCSS('reset') %>
	<% require themedCSS('typography') %>
	<% require themedCSS('form') %>
	<% require themedCSS('layout') %>
	<link rel="shortcut icon" href="$ThemeDir/images/favicon.ico" />
</head>
<% end_cached %>
```

**Comment by tractorcow** [comment]
I agree; We should fix this in the parser to stop treating it as a control, as you say. ;D

I wonder if we couldn't just make it `$BaseTag` and use template global provider for this? Maybe you could use it in your project in the mean time. :D

```php
class MyController {
	public function BaseTag() {
		return SSViewer::get_base_tag('');
	}
}
```

**Comment by jonom** [comment]
@HARVS1789UK side note: just be aware that the base tag includes the protocol, so if you allow connecting to your site on both http and https and cache the base tag you may get some undesirable results here.

**Comment by HARVS1789UK** [comment]
@tractorcow thanks, I am happy that I have it working for my current project now (although I might switch you your suggestion which looks a little cleaner), I just wanted to raise the issue so it could be resolved for others in future releases.

@jonom thanks for the heads up, I hadn't considered that but as it happens this site currently only serves content over http and when I eventually sort out an SSL cert I normally 301 redirect all http requests to their https equivalent
